### PR TITLE
j2objcXcode now dependency of j2objcBuild

### PR DESCRIPTION
**Pull request by brunobowden**
- This will update the Podfile when the j2objc project dependencies change

### CODE REVIEW

**Comment by brunobowden** [comment]
PTAL

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy [hunk 1]
```diff
@@ -277,7 +277,7 @@ class J2objcPlugin implements Plugin<Project> {
             }
             lateDependsOn(project, 'build', 'j2objcBuild')
 
-            // TODO: Where shall we fit this task in the plugin lifecycle?
+            // TODO: make this a dependency of j2objcBuild
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: Do you want to just go ahead and do this?

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy [hunk 2]
```diff
@@ -114,7 +125,23 @@ class XcodeTask extends DefaultTask {
     void xcodeConfig() {
         Utils.requireMacOSX('j2objcXcode task')
 
-        verifyXcodeArgs()
+        // TODO: figure out how to automatically run this on top-level project
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: what does this comment mean?

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy [hunk 3]
```diff
@@ -114,7 +125,23 @@ class XcodeTask extends DefaultTask {
...
+        // TODO: but must be disabled on all dependencies
+        if (!isTaskActive()) {
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: I agree with the goal where j2objcBuild depends on j2objcXcode.  But in that version, !isTaskActive is the way to bail out on this, and should not throw an error

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy (hunk 2 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: Ideally you'd like this error to apply on the root project, the same that builds the android app. You don't know if any project is the root or a dependency instead. I haven't figured out if there's an appropriate way to figure out when to show this error.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy (hunk 3 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: I was a little apprehensive about doing this... but going ahead and making the change. Also update the docs too.

**Comment by advayDev1** [comment]
Please add a comment to CHANGELOG.md also; LGTM when passes.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy [hunk 4]
```diff
@@ -53,6 +53,18 @@ class XcodeTask extends DefaultTask {
     @Input @Optional
     String getXcodeProjectDir() { return J2objcConfig.from(project).xcodeProjectDir }
 
+    boolean isTaskActive() { return getXcodeProjectDir() != null }
+
+    @Input
+    // List of all dependencies
+    List<PodspecDetails> getPodspecDependencies() {
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: class PodspecDetails must be modified as follows to be an input:
```
@EqualsAndHashCode
static class PodspecDetails implements Serializable {
...
private static final long serialVersionUID = 1L;
...
}
```

**Comment by brunobowden** [comment]
@advayDev1 - it also required implementing `writeObject()` and calling `defaultWriteObject()`. I added deserialization as well, so that I could test it end it end. Curiously, the `@EqualsAndHashCode` wasn't needed, so I wanted your opinion on that.

**Comment by advayDev1** [comment]
1.  Not throwing without EqualsAndHashCode makes sense.  However, I think you use it anyway: while serialization will work, you need it to make sure up-to-date checks are correct and even in Groovy, equality is referential unless overridden.

2.  I'm not understanding why writeObject/defaultWriteObject were needed for you.  I don't have them.  Instead of doing that, consider using @AutoExternalize: http://docs.groovy-lang.org/2.4.5/html/gapi/groovy/transform/AutoExternalize.html

**Comment by advayDev1** [comment]
Aha: you don't have a default no-arg constructor for PodspecDetails.  Ok this makes sense now

**Comment by brunobowden** [comment]
I expect that Gradle is comparing Serialized binary to Serialized binary, that way that can get consistent behaviour over time and across daemon runs - they can't keep the object around in memory. The writeDefaultObject is a necessary part of serialization.

**Comment by advayDev1** [comment]
I'd still suggest using @EqualsAndHashCode - we don't know what the implementation detail of Gradle is (I looked briefly, couldn't find it).  Understood about writeDefaultObject, the point of @AutoExternalizable is to do that for you.  (I'm still trying to figure out why I don't need AutoExternalizable nor implementations of Serializable - but that is an aside).

**Comment by brunobowden** [comment]
I've added back @EqualsAndHashCode. Since the serialization is working (and is tested), I'm fine with doing it the Java way with writeDefaultObject().

One concern I have is that getPodspecDependencies() is called 2-3 times during the gradle run. I'm surprised that this happens. Dependending on how large the dependency trees become, this may need caching.

I'll merge this once the build is green.

**Comment by advayDev1** [comment]
sg - LGTM

**Pull request merged**